2022/Q2/artifacts/pos-model/Staking.tla

```diff
@@ -0,0 +1,179 @@
+ *
+ * Simplifications:
+ *   - We assume only one validator to be delegated to.
+ *   - Rewards are not specified.
+ *   - Jailing is not specified.
+ *   - Slashing and evidence handling is not specified.
+ * 
+ * Manuel Bravo, 2022
+ *)
+EXTENDS Integers, Apalache, Staking_typedefs
+
+CONSTANTS
+    \* Set of all addresses on Cosmos.
+    \* @type: Set(ADDR);
+    UserAddrs
+
+VARIABLES
+    \* Coin balance for every Cosmos account.
+    \*
+    \* @type: BALANCE;
+    balanceOf,
+    \* Balance of unbonded coins that cannot be used during the bonding period.
+    \*
+    \* @type: BALANCE;
+    unbonded,
+    \* Coins that are delegated to Validator.
+    \*
+    \* @type: BALANCE;
+    delegated
+
+\* Variables that model transactions, not the state machine.
+VARIABLES    
+    \* The last executed transaction.
+    \*
+    \* @type: TX;
+    lastTx,
+    \* A serial number to assign unique ids to transactions
+    \* @type: Int;
+    nextTxId,
+    \* Counts the transaactions executed within an epoch
+    \* @type: Int;
+    txCounter,
+    \* A serial number used to identify epochs
+    \* @type: Int;
+    epoch,
+    \* Whether at least one transaction has failed
+    \* @type: Bool;
+    failed
+
+\* the maximum value in Cosmos
+MAX_UINT == 2^(256 - 60) - 1
+
+\* 1 billion coins in the initial supply
+INITIAL_SUPPLY == 10^(9+18)
+
+\* the number of coins the validator has staked
+INIT_VALIDATOR_STAKE == 1000000000000000000000
+
+\* tx per epoch
+TXS_PER_EPOCH == 10
+
+\* the validator account
+Validator == "validator"
+
+\* Initialize the balances
+Init ==
+    \* user{1,2,3} have 1M Cosmos coins, the validator has 1M - stake
+    /\ balanceOf = [ a \in UserAddrs |->
+        IF a /= "validator"
+        THEN INITIAL_SUPPLY
+        ELSE INITIAL_SUPPLY - INIT_VALIDATOR_STAKE
+       ]
+    /\ unbonded = [ a \in UserAddrs |-> 0 ]
+    /\ delegated = [
+        a \in UserAddrs |-> IF a /= "validator"
+        THEN 0
+        ELSE INIT_VALIDATOR_STAKE
+       ]
+    /\ nextTxId = 0
+    /\ epoch = 1
+    /\ txCounter = 0
+    /\ lastTx = [ id |-> 0, tag |-> "None", fail |-> FALSE ]
+    /\ failed = FALSE
+
+
+\* delegate `amount` coins to Validator
+Delegate(sender, amount) ==
+    LET fail ==
+        \/ amount < 0
+        \/ amount > balanceOf[sender]
+    IN
+    /\ lastTx' = [ id |-> nextTxId, tag |-> "delegate", fail |-> fail,
+                   sender |-> sender, toAddr |-> Validator, value |-> amount ]
+    /\ nextTxId' = nextTxId + 1
+    /\ failed' = (fail \/ failed)
+    /\  IF fail
+        THEN
+          UNCHANGED <<balanceOf, unbonded, delegated>>
+        ELSE
+          \* transaction succeeds
+          \* update the balance of 'sender'
+          /\ balanceOf' =
+                [ balanceOf EXCEPT ![sender] = @ - amount]
+          /\ delegated' = [ delegated EXCEPT ![sender] = @ + amount ]
+          /\ UNCHANGED unbonded
+
+
+\* unbond `amount` coins from Validator
+Unbond(sender, amount) ==
```

Comment:
the sender here is the validator in question?

Reply:
There is a single validator in the current model, so the validator in question does not need to be an argument of the function. The sender is the user that wants to unbond its token. We should discuss whether it makes sense to model multiple validators or not. I am not sure it will add much to the model.

Reply:
We'd need to test rebond... so probably need it